src/Smalot/PdfParser/PDFObject.php

```diff
@@ -127,100 +132,195 @@ public function getContent(): ?string
         return $this->content;
     }
 
-    public function cleanContent(string $content, string $char = 'X')
+    public function cleanContent(?string $content): string
```

Comment:
That is a backward compatibility break, because you removed a parameter from a public function. Can you tell me, how it was used until then?

Reply:
`cleanContent()` was (and still is) only called once by `getSectionsText()` in the entire of PdfParser. Previously, `cleanContent()` would behave... similarly to how it does in this PR. It would hide strings `()` and dictionaries `<<>>`, but also unnecessarily hide arrays `[]` as well. It would do this by replacing the entire item by a string of the same length made up of the `$char` character. So:

```
(Modificatio[ns] au \\(14\\) septembre 2009 ET 2010)Tj
```

... would become:

```
(________________________________________________)Tj
```

It would also unnecessarily hide `BDC` and `EMC` commands. But the hiding was not done for the same purpose as it is in this PR.

The current v2.7.0 version of `getSectionsText()` would accept the full, uncleaned content stream, then make a `cleanContent()` duplicate of it. It would then apply regexps to match "text blocks" and "'do' commands" from the 'cleaned' content. The offset values of those matches were then used to `substr()` the desired text from the *actual* content string.

However, during this process, the only time `q` and `Q` were acknowledged were if they were directly adjacent to the `BT` or `ET` of a text block. Likewise, the only time `BDC` and `EMC` commands were acknowledged were when they were *inside* a `BT ... ET` text block.

---

This PR changes the mission of `cleanContent()` from being a function that just makes a copy of the content with some items blanked out to make matching easier, to being a workhorse that **cleans** the content that actually gets worked on. It hides strings `()` and dictionaries `<<>>` like the old `cleanContent()` did, however once the rest of the document stream is correctly formatted, it *restores* these items so the content stream is complete. It ensures there is one and only one PDF operator command on each line.

`getSectionsText()` then, has been modified from its previous function of grabbing `BT ... ET` blocks by regexp, to stepping through the entire content stream, command by command, and *discarding* the commands we don't need.

Conclusion: I believe that the previous incarnation of `cleanContent()` didn't need to be a `public` function at all. It was only ever called once, and then only to make an incomplete copy for `getSectionsText()` to find offset values from. The output from it isn't used otherwise.

**However**: Now that `cleanContent()` returns a precisely formatted version of the *complete* content stream with one command per line, it now probably *does* have value as a public function, as users can call it to format arbitrary content streams for whatever reason they like.

Reply:
My only issue here is that changing parameters/return values of public functions is considered a backward compatibility break according to https://semver.org/#spec-item-8. But, it seems that the function in question is not really part of our public API. PDFObject is a mother class for other sub classes and not directly accessible be users (for most of the time).

On the other hand, developers might directly use these functions to tune their software, because they are public and are not marked or commented in any way to decline that form of usage.

Whatever we decide, please add `@internal For internal use only, not part of the public API` to the function header to signal developers to not rely on it in the future (About **internal** in PHP Doc: https://manual.phpdoc.org/HTMLSmartyConverter/HandS/phpDocumentor/tutorial_tags.internal.pkg.html).

I tend to go with your suggestion and keep your changes. However, I will add a short info about this change in the release notes. A release candidate first, before a stable reason might also be a good idea.

Reply:
Since `getSectionsText()` was the only function using `cleanContent()` perhaps I could *restore* the old function and then rename the one I've created to something new? It'll leave some old, unused code in the source, but it wouldn't be a backwards-incompatible change anymore. What do you think? I could call it something like... `formatContent()`?

Reply:
I restored `cleanContent()` to its previous version and renamed my function `formatContent()`. `cleanContent()` is now no longer called anywhere in PdfParser, but if any user relied on it, it's still there to use. It should probably be deprecated and then removed in future major versions.

Reply:
Is this satisfactorily resolved, @k00ni ?